src/lib.rs

```diff
@@ -1,6 +1,43 @@
 #![deny(missing_docs)]
 
-//! VMM Resource manager.
+//! VMM resource manager and allocator.
 
 mod interval_tree;
 pub use interval_tree::{IntervalTree, Range};
+
+/// Policy for resource allocation.
+#[derive(Copy, Clone, Debug)]
+pub enum AllocPolicy {
+    /// Default allocation policy.
+    Default,
+    /// Allocate from the first matched entry.
+    FirstMatch,
+}
+
+/// Struct to describe resource allocation constraints.
+#[derive(Copy, Clone, Debug)]
+pub struct Constraint {
+    /// Size to allocate.
+    pub size: u64,
+    /// Lower boundary for the allocated resource.
+    pub min: u64,
+    /// Upper boundary for the allocated resource.
+    pub max: u64,
+    /// Alignment for the allocated resource.
+    pub align: u64,
+    /// Resource allocation policy.
+    pub policy: AllocPolicy,
+}
+
+impl Constraint {
+    /// Create a new constraint object with default settings.
+    pub fn new(size: u64) -> Self {
```

Comment:
It seems this function needs more parameters?

Reply:
It's used to cover the most common usage scenario. For rare use cases, you may dirrectly construct a Constraint object or we may add some new interfaces for those cases.

Reply:
So what is the difference between this and `ResourceConstraint` defined in https://github.com/rust-vmm/vm-device/pull/9 ?

Reply:
It's two related abstractions. With Device::ResourceConstraint and vm_allocator::Constraint, there's still one piece missing of the whole puzzle. We need the vmm to implement a resource manager, which converts Device::ResourceConstraint to vm_allocator::Constraint.
1) vm_allocator is just a generic resource allocator,
2) vmm::ResourceManager needs to define type of resources and use vm_allocator to manage its resources.
It's still unclear whether vmm::ResourceManager should be VMM specific or we need another vmm_resource crate.

Reply:
> It's two related abstractions. With Device::ResourceConstraint and vm_allocator::Constraint, there's still one piece missing of the whole puzzle. We need the vmm to implement a resource manager, which converts Device::ResourceConstraint to vm_allocator::Constraint.
> 
Ok, this is the "glue" part between vm-device and vm-allocator.

> 1. vm_allocator is just a generic resource allocator,
> 2. vmm::ResourceManager needs to define type of resources and use vm_allocator to manage its resources.
>    It's still unclear whether vmm::ResourceManager should be VMM specific or we need another vmm_resource crate.

I think we also need a conversion from the vmm::ResourceManager::allocate_resources() return type to vm_device::Resource for registering devices.

Reply:
Yes, that's true. Interval tree just manages a set of IDs, the resource manager needs to define the specific resources the ID stands for.